snakemake/executors.py

```diff
@@ -2052,3 +2059,605 @@ def run_wrapper(
+        lookup = {}
+        for zone in zones:
+            request = self._compute_cli.machineTypes().list(
+                project=self.project, zone=zone["name"]
+            )
+            lookup[zone["name"]] = self._retry_request(request)["items"]
+
+        # Only keep those that are shared, use last zone as a base
+        machineTypes = {mt["name"]: mt for mt in lookup[zone["name"]]}
+        del lookup[zone["name"]]
+
+        # Update final list based on the remaining
+        to_remove = set()
+        for zone, types in lookup.items():
+            names = [x["name"] for x in types]
+            for machineType in list(machineTypes.keys()):
+                if machineType not in names:
+                    to_remove.add(machineType)
+
+        for machineType in to_remove:
+            del machineTypes[machineType]
+
+        # Alert the user of machineTypes available before filtering
+        # https://cloud.google.com/compute/docs/machine-types
+        logger.debug(
+            "found {} machine types across regions {} before filtering"
+            "to increase selection, define fewer regions".format(
+                len(machineTypes), self.regions
+            )
+        )
+
+        # First pass - eliminate anything that too low in cpu/memory
+        keepers = dict()
+        for name, machineType in machineTypes.items():
+            if machineType["guestCpus"] < cores or machineType["memoryMb"] < mem_mb:
+                continue
+            keepers[name] = machineType
+
+        # If a prefix is set, filter down to it
+        if self._machine_type_prefix:
+            machineTypes = keepers
+            keepers = dict()
+            for name, machineType in machineTypes.items():
+                if name.startswith(self._machine_type_prefix):
+                    keepers[name] = machineType
+
+        # If we don't have any contenders, workflow error
+        if not keepers:
+            raise WorkflowError(
+                "Machine prefix {prefix} is too strict, and there are no options "
+                "available.".format(prefix=self._machine_type_prefix)
+            )
+
+        # Now find (quasi) minimal to satisfy constraints
+        machineTypes = keepers
+        keepers = dict()
+
+        minCores = 1
+        minMem = 3.75
+        smallest = "n1-standard-1"
+
+        for name, machineType in machineTypes.items():
+            if machineType["guestCpus"] < minCores and machineType["memoryMb"] < minMem:
+                smallest = name
+                minCores = machineType["guestCpus"]
+                minMem = machineType["memoryMb"]
+
+        selected = machineTypes[smallest]
+        logger.debug(
+            "Selected machine type {}:{}".format(smallest, selected["description"])
+        )
+        virtualMachine = {
+            "machineType": smallest,
+            "labels": {"app": "snakemake"},
+            "bootDiskSizeGb": 100,  # default is likely 10
```

Comment:
Should this be disk_mb, maybe plus an offset?

Reply:
Is `disk_mb` an attribute that populates from Snakemake? Where does it come from?

Reply:
okay I'm going to assume `disk_mb` is a job resource, grab from there, default to 100 and add an offset of 50.

Reply:
or maybe just 10 is okay - I don't want to be single highhandedly responsible for people getting over-charged.

Reply:
So, disk_mb is a recommended resource (if needed), and people are able to set it via the command line with --default-resources (see the corresponding help test). The default is `max(2*input.size, 1000)`. You don't need to set any defaults in the executor yourself. Instead, require people to use --default-resources when they use the Google LS backend.

Reply:
When you say require - do you mean I should exit with error if the flag is not used? Or the default setting will handle setting it, and I should add to the documentation to encourage users to use the flag?

Reply:
Are you looking at old code? Locally I see:
```python
        virtualMachine = {
            "machineType": smallest,
            "labels": {"app": "snakemake"},
            "bootDiskSizeGb": disk_gb,
        }
```
and it's derived from
```python
        disk_mb = job.resources.get("disk_mb", 102400)  # 100 GB default

        # Convert mb to gb, add buffer of 50
        disk_gb = math.ceil(disk_mb / 1024) + 10
```